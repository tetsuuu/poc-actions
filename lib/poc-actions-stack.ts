import { Construct, Stack, StackProps }  from '@aws-cdk/core';
import { Group, Policy, PolicyStatement, ManagedPolicy, User } from '@aws-cdk/aws-iam';
// import * as AWS from 'aws-sdk';
// import IAM from 'aws-sdk/clients/iam';
// import { UpdateAccountPasswordPolicyRequest } from '@aws-sdk/client-iam-node';

const admins = 'testAdminGroup';
const adminUsers = [
  'demo01',
  // 'User02',
  // 'User03'
];

const developers = 'testDevGroup';
const devUsers = [
  'demo02',
  'demo03'
];

export class CdkIamStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // // import aws-sdk for JavaScript `npm install aws-sdk`
    // const accountPasswordPolicy = new AWS.IAM({});
    // accountPasswordPolicy.UpdateAccountPasswordPolicyRequest({
    //   "MinimumPasswordLength": 8,
    //   "RequireSymbols": false,
    //   "RequireNumbers": true,
    //   "RequireUppercaseCharacters": true,
    //   "RequireLowercaseCharacters": true,
    //   "AllowUsersToChangePassword": true,
    //   "ExpirePasswords": false
    // });
    
    //アカウントのパスワードポリシーを見に行く権限
    const getAccountPassword = new PolicyStatement({
        resources: ["*"],
        actions: [
          "iam:GetAccountPasswordPolicy"
        ]
    });

    //IAMユーザーのパスワード変更権限
    const changePassword = new PolicyStatement({
      resources: ["arn:aws:iam::account-id-without-hyphens:user/${aws:username}"],
      actions: [
        "iam:ChangePassword",
      ]
    });

    //IAMロールへのアクセス権限
    const iamPassRoleAccess = new PolicyStatement({
      resources: ["*"],
      actions: [
        "iam:Get*",
        "iam:List*",
        "iam:PassRole"
        ],
    });

    //AWS管理ポリシー
    const adminPolicy = ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess');
    const powerUserPolicy = ManagedPolicy.fromAwsManagedPolicyName('PowerUserAccess');

    //Devユーザー用のポリシー
    const devPolicy = new Policy(this, 'iamPassRoleAccess', { 
      policyName: "iamPassRoleAccess",
      statements: [iamPassRoleAccess],
    });

    //共通のポリシー
    const commonPolicy = new Policy(this, 'changePassword', { 
      policyName: "changePassword",
      statements: [changePassword, getAccountPassword],
    });

    //Adminユーザーグループ
    const adminGroup = new Group(this, admins, { groupName: admins });
    adminGroup.addManagedPolicy(adminPolicy);
    adminGroup.attachInlinePolicy(commonPolicy);

    //Devユーザーグループ
    const devGroup = new Group(this, developers, { groupName: developers });
    devGroup.addManagedPolicy(powerUserPolicy);
    devGroup.attachInlinePolicy(devPolicy);
    devGroup.attachInlinePolicy(commonPolicy);

    //ユーザーが作成されない…
    adminUsers.forEach(adminUser => {
      new User(this, adminUser, {
        userName: adminUser,
        groups: [adminGroup],
      });
    });

    devUsers.forEach(devUser => {
      new User(this, devUser, {
        userName: devUser,
        groups: [devGroup]
      });
    });
  }
}
