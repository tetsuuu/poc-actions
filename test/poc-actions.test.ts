import '@aws-cdk/assert/jest'
import { SynthUtils } from '@aws-cdk/assert'
import { Stack, App } from '@aws-cdk/core'  
 
import { CdkIamStack } from '../lib/poc-actions-stack'
 
test('IAM users fine', () => {
  const app = new App();
  const stack = new CdkIamStack(app, 'usersFineGrainedTestStack');
  expect(stack).toHaveResource('AWS::IAM::User', {
    UserName: 'demo01',
    Groups: [
      {
        "Ref": "testAdminGroupA356E014"
      }
    ]
  })
})
   
test('IAM group fine', () => {
  const app = new App();
  const stack = new CdkIamStack(app, 'groupFineGrainedTestStack');
  expect(stack).toHaveResource('AWS::IAM::Group', {
    GroupName: 'testAdminGroup',
    "ManagedPolicyArns": [
      {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition"
            },
            ":iam::aws:policy/AdministratorAccess"
          ]
        ]
      }
    ]
  })
})

test('IAM policy fine', () => {
  const app = new App();
  const stack = new CdkIamStack(app, 'policyFineGrainedTestStack');
  expect(stack).toHaveResource('AWS::IAM::Policy', {
    PolicyName: 'iamPassRoleAccess',
    Groups: [
        // {"Ref": "testAdminGroupA356E014"},
        {"Ref": "testDevGroup93FABFEE"}
    ],
    PolicyDocument: {
      "Statement": [
        {
          "Action": [
              "iam:Get*",
              "iam:List*",
              "iam:PassRole",
            ],
          "Effect": "Allow",
          "Resource": "*",
        },
      ],
      "Version": "2012-10-17",
    }
  })
})

test('IAM Resource Stack', () => {
  const app = new App();
  const stack = new CdkIamStack(app, 'IAMTestStack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
