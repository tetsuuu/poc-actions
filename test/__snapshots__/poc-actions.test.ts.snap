// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IAM Resource Stack 1`] = `
Object {
  "Resources": Object {
    "changePassword30BB1E33": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testAdminGroupA356E014",
          },
          Object {
            "Ref": "testDevGroup93FABFEE",
          },
        ],
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iam:ChangePassword",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::account-id-without-hyphens:user/\${aws:username}",
            },
            Object {
              "Action": "iam:GetAccountPasswordPolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "changePassword",
      },
      "Type": "AWS::IAM::Policy",
    },
    "demo01B48C8159": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testAdminGroupA356E014",
          },
        ],
        "UserName": "demo01",
      },
      "Type": "AWS::IAM::User",
    },
    "demo02CD55601F": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testDevGroup93FABFEE",
          },
        ],
        "UserName": "demo02",
      },
      "Type": "AWS::IAM::User",
    },
    "demo03EEE263E8": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testDevGroup93FABFEE",
          },
        ],
        "UserName": "demo03",
      },
      "Type": "AWS::IAM::User",
    },
    "iamPassRoleAccessD553F35C": Object {
      "Properties": Object {
        "Groups": Array [
          Object {
            "Ref": "testDevGroup93FABFEE",
          },
        ],
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "iam:Get*",
                "iam:List*",
                "iam:PassRole",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "iamPassRoleAccess",
      },
      "Type": "AWS::IAM::Policy",
    },
    "testAdminGroupA356E014": Object {
      "Properties": Object {
        "GroupName": "testAdminGroup",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AdministratorAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "testDevGroup93FABFEE": Object {
      "Properties": Object {
        "GroupName": "testDevGroup",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/PowerUserAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
  },
}
`;
